src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,558 @@ namespace Microsoft.CodeAnalysis.CSharp
-            loweredLeft = ConvertConcatExprToString(loweredLeft);
-            loweredRight = ConvertConcatExprToString(loweredRight);
+            return CreateStringConcat(syntax, arguments);
+        }
 
-            Debug.Assert(loweredLeft.Type is { } && (loweredLeft.Type.IsStringType() || loweredLeft.Type.IsErrorType()) || loweredLeft.ConstantValueOpt?.IsNull == true);
-            Debug.Assert(loweredRight.Type is { } && (loweredRight.Type.IsStringType() || loweredRight.Type.IsErrorType()) || loweredRight.ConstantValueOpt?.IsNull == true);
+        private BoundExpression VisitStringConcatenation(BoundBinaryOperator originalOperator)
+        {
+            Debug.Assert(IsBinaryStringConcatenation(originalOperator));
 
-            // try fold two args without flattening.
-            var folded = TryFoldTwoConcatOperands(loweredLeft, loweredRight);
-            if (folded != null)
+            if (_inExpressionLambda)
             {
-                return folded;
+                // If this is an expression tree, we can't optimize anything. Just do a standard visit and return.
+                return RewriteStringConcatInExpressionLambda(originalOperator);
             }
 
-            // flatten and merge -  ( expr1 + "A" ) + ("B" + expr2) ===> (expr1 + "AB" + expr2)
-            ArrayBuilder<BoundExpression> leftFlattened = ArrayBuilder<BoundExpression>.GetInstance();
-            ArrayBuilder<BoundExpression> rightFlattened = ArrayBuilder<BoundExpression>.GetInstance();
+            // We'll walk the children in a depth-first order, pull all the arguments out, and then visit them. We'll fold any constant arguments as
+            // we go, pulling them all into a string literal.
+            CollectAndVisitConcatArguments(originalOperator, visitedCompoundAssignmentLeftRead: null, out var arguments);
 
-            FlattenConcatArg(loweredLeft, leftFlattened);
-            FlattenConcatArg(loweredRight, rightFlattened);
+            return CreateStringConcat(originalOperator.Syntax, arguments);
+        }
 
-            if (leftFlattened.Any() && rightFlattened.Any())
+        /// <summary>
+        /// Produces a new string.Concat call in the most efficient manner for the given arguments. It is expected that the arguments are already visited, and the following optimizations
+        /// have been done:
+        /// <list type="number">
+        /// <item>Any consecutive constant strings or chars have been folded.</item>
+        /// <item>Any nested string.Concat calls have had their arguments deconstructed into <paramref name="visitedArguments"/>.</item>
+        /// </list>
+        /// It is not valid to call this method inside an expression tree; that should be handled by a standard recursive rewrite.
+        /// </summary>
+        private BoundExpression CreateStringConcat(SyntaxNode originalSyntax, ArrayBuilder<BoundExpression> visitedArguments)
+        {
+            Debug.Assert(!_inExpressionLambda);
```

Comment:
nit: Consider asserting that the arguments have either string, char or object type.